File: src/main/java/spring/springserver/global/exception/exception/ApplicationException.java
package spring.springserver.global.exception.exception;

import spring.springserver.global.exception.status_code.StatusCode;
import lombok.Getter;

@Getter
public class ApplicationException extends RuntimeException {

  private final StatusCode statusCode;

  public ApplicationException(StatusCode statusCode) {

    super(statusCode.getMessage());
    this.statusCode = statusCode;
  }

  public ApplicationException(StatusCode statusCode,
                              Throwable throwable) {

    super(statusCode.getMessage(), throwable);
    this.statusCode = statusCode;
  }

  public ApplicationException(StatusCode statusCode,
                              String message) {

    super(message);
    this.statusCode = statusCode;
  }

  public static ApplicationException of(StatusCode statusCode) {

    return new ApplicationException(statusCode);
  }

  public static ApplicationException of(StatusCode statusCode,
                                        Throwable throwable) {

    return new ApplicationException(
            statusCode,
            throwable
    );
  }

  public static ApplicationException of(StatusCode statusCode,
                                        String customMessage) {

    return new ApplicationException(
            statusCode,
            customMessage
    );
  }
}

File: src/main/java/spring/springserver/global/exception/exception/ApplicationException.kt
package spring.springserver.global.exception.exception

import spring.springserver.global.exception.status_code.StatusCode

class ApplicationException(
    val statusCode: StatusCode,
    message: String = statusCode.getMessage(),
    throwable: Throwable? = null
) : RuntimeException(message, throwable) {

    constructor(statusCode: StatusCode, throwable: Throwable) : this(
        statusCode = statusCode,
        message = statusCode.getMessage(),
        throwable = throwable
    )

    companion object {

        fun of(
            statusCode: StatusCode
        ): ApplicationException {

            return ApplicationException(statusCode)
        }

        fun of(
            statusCode: StatusCode,
            throwable: Throwable
        ): ApplicationException {

            return ApplicationException(statusCode, throwable)
        }

        fun of(
            statusCode: StatusCode,
            customMessage: String
        ): ApplicationException {

            return ApplicationException(statusCode, customMessage)
        }
    }
}
